src/Sparrow.Server/Allocators.Pool.FixedSizeThreadAffine.cs

```diff
@@ -0,0 +1,371 @@
+
+        // PERF: This should be devirtualized. 
+        private IAllocatorComposer<Pointer> _internalAllocator;
+
+        private Container[] _container;
+
+        private struct Container
+        {
+            public IntPtr Block1;
+            public IntPtr Block2;
+            public IntPtr Block3;
+            public IntPtr Block4;
+        }
+
+        public long TotalAllocated
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get;
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private set;
+        }
+
+        public long Allocated
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get;
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private set;
+        }
+        public long InUse
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get;
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private set;
+        }
+
+        public void Initialize(ref FixedSizeThreadAffinePoolAllocator<TOptions> allocator)
+        {
+            allocator._container = new Container[(int)allocator._options.Workload]; // PERF: This should be a constant.       
+
+            allocator.TotalAllocated = 0;
+            allocator.Allocated = 0;
+            allocator.InUse = 0;
+        }
+
+        public void Configure<TConfig>(ref FixedSizeThreadAffinePoolAllocator<TOptions> allocator, ref TConfig configuration) where TConfig : struct, IAllocatorOptions
+        {
+            if (!typeof(TOptions).GetTypeInfo().IsAssignableFrom(typeof(TConfig)))
+                throw new NotSupportedException($"{nameof(TConfig)} is not compatible with {nameof(TOptions)}");
+
+            // This cast will get evicted by the JIT. 
+            allocator._options = (TOptions)(object)configuration;
+            allocator._internalAllocator = allocator._options.CreateAllocator();
+
+            // We are not going to support non thread safe allocator as 'backing allocators'. 
+            if (!allocator._internalAllocator.IsThreadSafe)
+                throw new NotSupportedException($"The internal allocator '{allocator._internalAllocator.GetType().Name}' is not thread-safe, therefore not supported to be used by this allocator.");
+
+            if (allocator._options.ItemsPerLane < 1)
+                throw new ArgumentOutOfRangeException($"{nameof(allocator._options.ItemsPerLane)} cannot be smaller than 1.");
+            if (allocator._options.ItemsPerLane > 4)
+                throw new ArgumentOutOfRangeException($"{nameof(allocator._options.ItemsPerLane)} cannot be bigger than 4.");            
+        }
+
+        public Pointer Allocate(ref FixedSizeThreadAffinePoolAllocator<TOptions> allocator, int size)
+        {
+            if (allocator._options.AcceptOnlyBlocks && size != allocator._options.BlockSize)
+                throw new InvalidOperationException(
```

Comment:
Move to dedicated method?

Reply:
No need now. that is Pre-2.0 JIT behavior. Until we are sure that we are only using 2.0 JIT we need to keep it, but this is going to be executed on hopefully 3.0 JIT so no need to do that anymore.